Implement grouping functionality

CMDRyan opened the pull request:
@Steven-Chan 

UI changes
- Profile icon has been replaced with hamburger icon
- Hamburger icon will trigger curtain menu which includes groupby dropdown box

Added features
- Library can now handle groupBy function call and will return data to client in an array format
- GroupBy function has been added
   - By machine type
   - By Project name

To-Do:
- Align number designs

Steven-Chan commented on src/getData.js:
```diff
@@ -12,16 +12,25 @@ module.exports.getProjectIDs = function getProjectIDs(requestedGroupID, token) {
-      return ids;
+      const projectList = [];
+      projects.map((project) => {
+        Project = {
+          id: project.id,
+          name: project.name,
+        };
+        projectList.push(Project);
```
Bad code style. Missing "const". Variable name should be camelcase starts with small letters.

Steven-Chan replied:
I just found this style also appears somewhere else, pls help fix them altogether.

CMDRyan commented:
@Steven-Chan 

- Added const to object decorations
- Refactored groupBy "none", following the same style as other criteria
- Refactored group titles html building method using `Template` element
- Modified title css to compatible with the `Template` element

CMDRyan commented:
@Steven-Chan

Changes to UI:

- Changed feedback for cards to colour changing (white to light gray)
- Removed curtain navigation menu

<img width="1440" alt="Screenshot 2021-09-05 at 9 59 23 PM" src="https://user-images.githubusercontent.com/85364009/132129812-a8961ddb-0abe-4af5-bda2-040498430b6c.png">

<img width="378" alt="Screenshot 2021-09-05 at 10 01 01 PM" src="https://user-images.githubusercontent.com/85364009/132129822-e9c056f5-1021-4d57-8f35-e34f35377c4e.png">

Steven-Chan commented on src/views/partials/jobCard.ejs:
```diff
@@ -1,17 +1,35 @@
+  <div class="<%=job[current].status%>-card flex flex-col mt-0.5 mx-0.5 sm:p-5 sm:h-26 p-1 min-h-0 w-auto border-solid border-1 text-gray-400 bg-gray-50 transition duration-100 hover:bg-gray-400">
+
+    <div class="flex flex-row">
+      <div class="flex flex-grow-0">Project:</div>
```
Also you only specify the flex-grow, but not flex-shrink. Most cases I think we don't specify flex-grow or flex-shrink separately but use flex-1 directly. If the element has a fixed size, i think you better want to give it a fix width instead.

Steven-Chan commented on src/views/partials/singleJobCard.ejs:
```diff
@@ -1,17 +1,35 @@
-      tags: <span class="text-gray-400 relative">
-        <%= job.tag_list %>
-      </span>
+  <a href="<%=job.web_url%>" target="_blank">
```
I just realize there is a `jobCard.js` and `singleJobCard.js`. Can you help refactor and make them one single template file.

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -63,5 +94,73 @@
+      $(`#`+$(this).attr("id")).html(htmlArr)
+    }
+  </script>
+  <template>
+    <div class="flex flex-row">
+      <div class="flex flex-grow"></div>
+      <div class="flex flex-grow-0"></div>
+      <div class="flex flex-grow-0"></div>
+      <div class="flex flex-grow"></div>
+    </div>
+  </template>
```
What is this?

Steven-Chan commented on src/views/pages/index.ejs:
```diff
@@ -63,5 +94,73 @@
+        var titleTemplate = $("Template")[0]
+        var titleItem = titleTemplate.content.cloneNode(true)
+        var titleDiv = titleItem.querySelectorAll("div")
+        titleDiv[0].classList.add(`${this.id}-group-title`)
+        titleDiv[2].innerHTML = groupName
+        titleDiv[3].classList.add(`${this.id}-circle-sm`)
+        titleDiv[3].innerHTML = groupJobs.length
```
Don't use index to access the element.

Steven-Chan commented:
@CMDRyan  Pls also include screenshot with card.

CMDRyan commented:
Full screen: (hovering on one card):
<img width="1335" alt="Screenshot 2021-09-08 at 11 48 16 AM" src="https://user-images.githubusercontent.com/85364009/132443508-a25a0abf-c3f7-4fed-88ef-75c105aa1a07.png">

Mobile size (grouped by):
<img width="499" alt="Screenshot 2021-09-08 at 11 48 59 AM" src="https://user-images.githubusercontent.com/85364009/132443550-234fbfc8-b066-47cc-a42d-0b907c4ebad2.png">

Steven-Chan reviewed (changes requested):
@CMDRyan 
- The section header hight needs larger, it is too tight now
- The job count in section header is not aligned
- The job card spacing looks different in narrow screen and wide screen

CMDRyan commented:
@Steven-Chan 

- The cards rendering are refactored, and the extra jobCards.js is removed
- The header section padding has been increased
- Job card spacing in mobile view has been fixed
- Template function no longer uses index method to change element

CMDRyan commented:
Margin fixed for mobile view: 

<img width="572" alt="Screenshot 2021-09-09 at 11 33 21 AM" src="https://user-images.githubusercontent.com/85364009/132618124-f2b58134-4759-4c2d-a76d-e4263d9f8511.png">

CMDRyan commented:
Grouping header paddings and number fix:
<img width="394" alt="Screenshot 2021-09-09 at 1 03 18 PM" src="https://user-images.githubusercontent.com/85364009/132625852-2050be47-04f1-4383-bfbd-d5031ac86012.png">

Steven-Chan commented on src/app.js:
```diff
@@ -108,10 +108,46 @@ app.get("/groups/:id/jobs", function (req, res) {
+      const cardTemplate = ejs.compile(
+        read("src/views/partials/singleJobCard.ejs", "utf-8")
+      );
+      createdJobs.map((job)=>{
+        const jobObj = {
+          name: job.project_name,
+          tags: job.tag_list,
+          html: cardTemplate({job: job})
+        }
+      createdCards.push(jobObj)
+      })
+      pendingJobs.map((job)=>{
+        const jobObj = {
+          name: job.project_name,
+          tags: job.tag_list,
+          html: cardTemplate({job: job})
+        }
+      pendingCards.push(jobObj)
+      })
+      runningJobs.map((job)=>{
+        const jobObj = {
+          name: job.project_name,
+          tags: job.tag_list,
+          html: cardTemplate({job: job})
+        }
+      runningCards.push(jobObj)
```
This is not the correct way to use map. You transform the value of an array with map and the output of map is the array of new value already.

Steven-Chan commented on src/app.js:
```diff
@@ -108,10 +108,46 @@ app.get("/groups/:id/jobs", function (req, res) {
+      const createdJobs = data.filter((data) => data.status === "created")
+      const pendingJobs = data.filter((data) => data.status === "pending")
+      const runningJobs = data.filter((data) => data.status === "running")
+      const cardTemplate = ejs.compile(
+        read("src/views/partials/singleJobCard.ejs", "utf-8")
+      );
```
The template does not change in every request, so we should load it globally once.

CMDRyan commented:
@Steven-Chan 
- Card template has been made global
- Corrected logic of array mappings
